File: chap2/sec2.1/FindClosestInterval.m
function m=FindClosestInterval(x,y)
% FINDCLOSESTINTERVAL find closest interval in a vector to a number
%   [m1,m2]=FindClosestInterval(x,y) finds in the vector y the
%   interval which contains the value x, or if there is no such
%   interval, the one which is closest to x, and returns the lower
%   index m of this interval in y.

n=length(y); i=1;
while i<=n & x>y(i), i=i+1; end;
if i==1, m=1; elseif i==n+1, m=n-1; else m=i-1; end;

File: chap2/sec2.1/FindClosestInterval.py
def findClosestInterval(x,y):
    """
    Finds the closest interval in a vector to a number.

    Parameters
    ----------
    x : float
        The value for which to find the closest interval.
    y : list or numpy.ndarray
        A sorted vector defining the intervals.

    Returns
    -------
    int
        The lower index `m` of the interval in `y` that contains `x`.
        If `x` is outside the range of `y`, returns the closest interval.
    """
    n = len(y)
    i = 0
    while i < n and x > y[i]:
        i += 1
    if i == 0:
        return 0
    elif i == n:
        return n-1
    else:
        return i-1